### ISSUE TITLE
[Feature Request]: Integration of PySpark 

### ISSUE THREAD

**Issue by saxenam06** [visual-layer/fastdup#263]
### Feature Name

Integration of PySpark

### Feature Description

Hi, Many Thanks for open sourcing the interesting work. I am very new to this repo so my request could be naive. 
I was wondering why not leverage Pyspark for Faster training and Inference!
Also, I was wondering about how the scale was actually done in this repo without using Pyspark. 
 I am slightly stressing over Pyspark as I have seen many image-based applications successfully scale using it.
Could this be a future extension or my understanding is wrong and it is out of scope?
Many Thanks

**Comment by dbickson** [comment]
Hi @saxenam06 we need to understand better your needs. Pyspark does not natively support images :-)
The open source is limited for 1M images but we have some internal demos on up to 1B images. We typically prefer a single multicore machine to speed up the compute. Often the bottleneck is in getting the images to the compute node and not the actual compute which we do very efficiently in 2msec per core.
How many images do you have? We would love to help to make sure you are succesful.

**Comment by saxenam06** [comment]
Thank you So much for your quick response. 
My use case is to do image search semantically similar to a/many text queries. This may involve first of all creating Captions/Embedding Vectors using a model for e.g., BLIP for say 1M images on an Open-Source Autonomous Driving Perception dataset. Post this, use the text to query the topmost similar images. Once this is successful, I would like to create more clusters/classes of the images as per unique features that were not present in the original annotated metadata (which had only basic classes like sunny, cloudy but not Indepth like "Car on an intersection with a Pedestrian Infront on a rainy afternoon"). I already implemented this application using Pyspark on Google Colab/PySpark/PandasUdf using below example from Databricks. However, I could only achieve this application working on only 200 images in ~5 minutes processing entirely on Colab CPU. Though, I can move to Colab GPU, but I want to scale the computing resources step by step making use of the best software tools around first rather than simply leveraging more resources. Any further insights from your experience are deeply appreciated. I still did not understand the key technology in this repo using which you could scale so much to 1B images.    
https://docs.databricks.com/en/_extras/notebooks/source/deep-learning/dist-img-infer-2-pandas-udf.html.

**Comment by dbickson** [comment]
HI @saxenam06 we will be happy to help. The default colab node has 2 cores only and is very very very weak compute node. Using fastdup the speed of the compute is related to the number of cores. So if you have 32 cores it will run roughly x16 times faster. We are adding now similar search capabilities and would love to work with you if you want to share your use case. Can we set up a short zoom to share and discuss?

**Comment by saxenam06** [comment]
Hi @dbickson, Many thanks for your offer to help and your interest. After working on the colab, I learned that atleast by using Pyspark I was able to create vector embeddings for 200 images and retrieve top 3 matching images using a user text. The same was not possible without using distributed inference capabilities from spark. Therefore, I am convinced that Spark is one of an important piece of the solution. Now, if I like to scale I would like to do that on top of this may be by using AWS EMR/EMR server less based piepelines which scales well the spark code. For me sharing the use case is not a problem: here the link https://github.com/saxenam06/Image_retrieval_from_UserText_BLIP with some initial results using Pyspark on Colab.
